logdemux.1
.\" This document is in the public domain.
.\" vim: fdm=marker
.
.\" FRONT MATTER {{{
.Dd Oct 10, 2012
.Os
.Dt LOGDEMUX 1
.
.Sh NAME
.Nm logdemux
.Nd Logfile demultiplexer
.\" FRONT MATTER }}}
.
.\" SYNOPSIS {{{
.Sh SYNOPSIS
.Nm
.Ar rules
.Ar prefix
.\" SYNOPSIS }}}
.
.\" DESCRIPTION {{{
.Sh DESCRIPTION
.Nm
stores lines read from the standard input into different
files based on matching those lines with user-specified
regular expressions.
.
.Bl -tag -width "xx"
.It Ar rules
configuration file
.It Ar prefix
substituted into
.Li sink
values
.Pq Li \&%P
.El
.Pp
.
.Ss Rules
.
The filtering rules are given in an INI-formatted file, see
Iniphile for details on general syntax.
.Pp
Configuration must contain a
.Ql rules
section, with
.Ql order
property.
Its value is expected to list rule section names in the order they
should be used.
.Pp
Each rule section defines a rule with properties
.Ql match ,
.Ql sink ,
and optionally
.Ql final .
.Pp
.
.Bl -tag -width "xx"
.It Li match
Perl-compatible regular expression.
.
.It Li sink
Log file path.
.Li \&%P
is replaced with the
.Ar prefix
argument.
.Li \&%D
is replaced with current date in the
.Li YYYY-MM-DD
format.
.
.It Li final
Boolean.
.Li true , yes , on , 1
are true, anything else is false.
.El
.
.Ss Operation
Runtime behavior conforms to this pseudocode:
.Bd -literal
  for each line of input
    for each rule of rules.order
      if rule.match line
        write line to rule.sink
        if rule.final
          next line
.Ed
.
.\" DESCRIPTION }}}
.\" .Sh IMPLEMENTATION NOTES
.\" ENVIRONMENT {{{
.Sh ENVIRONMENT
None.
.\" ENVIRONMENT }}}
.\" .Sh FILES
.\" EXAMPLES {{{
.Sh EXAMPLES
This configures the Apache web server to log requests into
.Pa logs/%D-access_log ,
404's are duplicated into
.Pa logs/%D-404_log .
.Pp
ErrorLog is written into
.Pa logs/%D-error_log ,
except mod_fcgid, which goes into
.Pa logs/%D-mod_fcgid_log ,
except the FCGIWrapper's
.Dv stderr ,
which goes into
.Pa logs/%D-php_log .
.Pp
Note that
.Ql \&%D
is replaced with current date; see
.Sx DESCRIPTION .
.
.Pp
.Pa httpd.conf :
.
.Bd -literal -offset "xx"
LogFormat "%{%Y-%m-%d %H:%M:%S}t D=%D h=%h u=%u s=%>s b=%b r=\e"%r\e""
TransferLog "|/usr/local/bin/logdemux conf/logdemux.access logs/"
ErrorLog "|/usr/local/bin/logdemux conf/logdemux.error logs/"
.Ed
.
.Pp
.Pa logdemux.access :
.
.Bd -literal -offset "xx"
[rules]
order = http-404 default

[http-404]
match = \e<s=404\e>
sink = %P%D-404_log

[default]
match = .
sink = %P%D-access_log
.Ed
.
.Pp
.Pa logdemux.error :
.
.Bd -literal -offset "xx"
[rules]
order = php mod_fcgid default

[php]
match = [ ]mod_fcgid: stderr:[ ]
sink = %P%D-php_log
final = 1

[mod_fcgid]
match = [ ]mod_fcgid:[ ]
sink = %P%D-mod_fcgid_log
final = 1

[default]
match = .
sink = %P%D-error_log
.Ed
.
.\" EXAMPLES }}}
.\" DIAGNOSTICS {{{
.Sh DIAGNOSTICS
.Ss Exit code
.Bl -tag -width "xxxx" -compact
.It 0
successful operation
.It 66
the input file could not be open
.It 65
syntax error in input file
.It 64
wrong (number of) arguments
.El
.\" DIAGNOSTICS }}}
.\" .Sh COMPATIBILITY
.\" .Sh SEE ALSO {{{
.Sh SEE ALSO
.Lk https://github.com/roman-neuhauser/iniphile/ Iniphile
.\" .Sh SEE ALSO }}}
.\" .Sh STANDARDS
.\" .Sh HISTORY
.\" AUTHORS {{{
.Sh AUTHORS
.
.Nm
and this manual page are written by
.An Roman Neuhauser Aq Mt neuhauser@sigpipe.cz .
.Pp
See
.Lk https://github.com/roman-neuhauser/logdemux/ .
.\" AUTHORS }}}
.\" .Sh BUGS


logdemux.cpp
// Copyright (c) 2012 Roman Neuhauser
// Distributed under the MIT license (see LICENSE file)
// vim: sw=2 sts=2 ts=2 fdm=marker cms=\ //\ %s

// shared_ptr
#include <memory>

#include <iostream>
#include <istream>
#include <ostream>
#include <fstream>
#include <string>
#include <vector>

#define EX_OK           0       /* successful termination */
#define EX_USAGE        64      /* command line usage error */
#define EX_DATAERR      65      /* data format error */
#define EX_NOINPUT      66      /* cannot open input */
#define EX_SOFTWARE     70      /* internal software error */

#include "boost/ref.hpp"
#include "boost/foreach.hpp"
#include "boost/format.hpp"
#include "boost/date_time/gregorian/gregorian.hpp"
#include "boost/regex.hpp"

#include "iniphile/input.hpp"
#include "iniphile/ast.hpp"
#include "iniphile/output.hpp"

#define foreach BOOST_FOREACH

using std::shared_ptr;

using std::exception;

using std::cin;
using std::cerr;
using std::endl;
using std::ios;
using std::getline;
using std::ostream;
using std::ifstream;
using std::ofstream;
using std::string;
using std::vector;

using boost::cref;
using boost::format;
using namespace boost::gregorian;
using boost::regex;
using boost::smatch;

namespace
{

struct expand_sink // {{{
{
  expand_sink(string const &prefix, date const &today)
  : prefix(prefix)
  , today(today)
  {}
private:
  string const &prefix;
  date const &today;
public:
  string
  operator()(smatch const &expando) const
  {
    auto sexpando = expando.str();
    if (sexpando == "%D") return to_iso_extended_string(today);
    if (sexpando == "%P") return prefix;
    return "LOGDEMUX-BUG";
  }
}; // }}}

struct logdemux_error
: exception {};

struct missing_rules_order
: logdemux_error {};

class rule
// {{{
{
public:
  rule(string const &prefix, string const &match, string const &sink, bool final, ostream &diag) // {{{
  : sink(sink)
  , prefix(prefix)
  , final(final)
  , pat(match, regex::perl)
  , opened_on()
  , os()
  , diag(diag)
  {
  } // }}}
  bool
  handle(date const &today, string const &line) // {{{
  {
    if (!regex_search(line, pat))
      return false;
    if (opened_on != today)
      reopen(today);
    os << line << endl;
    return final;
  } // }}}
private:
  string const sink;
  string const prefix;
  bool final;
  regex pat;
  date opened_on;
  ofstream os;
  ostream &diag;

  void
  reopen(date const &today) // {{{
  {
    if (os.is_open()) os.close();
    os.clear();
    auto fname = expand(sink, today);
    os.open(fname.c_str(), ios::app | ios::binary);
    if (os.fail())
      diag << "failed to open " << fname << endl;
    opened_on = today;
  } // }}}
  string
  expand(string const &fmt, date const &d) const // {{{
  {
    return regex_replace(
      fmt
    , regex("%[DP]\\>")
    , cref(expand_sink(prefix, d))
    );
  } // }}}
}; // }}}

class ruleset
// {{{
{
public:
  ruleset(iniphile::ast::node const &ini, string const &prefix, ostream &diag) // {{{
  {
    auto order = iniphile::get(ini, "rules.order", vector<string>());
    if (order.empty())
      throw missing_rules_order();

    foreach (auto const &rname, order)
      rules.push_back(shared_ptr<rule>(new rule(
        prefix
      , iniphile::get(ini, rname + ".match", string("")) 
      , iniphile::get(ini, rname + ".sink", string(""))  
      , iniphile::get(ini, rname + ".final", false)      
      , diag
      )));
  } // }}}
  void
  handle(date const &now, string const &line) // {{{
  {
    foreach (auto &r, rules)
      if (r->handle(now, line))
        break;
  } // }}}
private:
  vector<shared_ptr<rule>> rules;
}; // }}}

date
today() // {{{
{
  return day_clock::local_day();
} // }}}

template<class Fmt>
int
complain(int exitcode, Fmt msg) // {{{
{
  cerr << msg << endl;
  return exitcode;
} // }}}

string
tailname(string const &path) // {{{
{
  return regex_replace(
    path
  , regex("^(.*/)?([^/]+)(.exe)?$", regex::perl)
  , "$2"
  );
} // }}}

}

int
main(int argc, char **argv)
{
  string self = argc ? tailname(argv[0]) : "logdemux";

  if (argc < 3)
    return complain(
      EX_USAGE
    , format("usage: %1% rules prefix") % self
    );

  string inipath = argv[1];
  string prefix = argv[2];
  ifstream input;
  input.open(inipath);
  if (input.fail())
    return complain(
      EX_NOINPUT
    , format("%1%: rules file '%2%' missing") % self % inipath
    );

  auto ini = iniphile::parse(inipath, input, cerr);
  input.close();
  if (!ini)
    return complain(
      EX_DATAERR
    , format("%1%: rules file '%2%' broken") % self % inipath
    );

  auto cfg(iniphile::normalize(*ini));

  try {
    ruleset rules(cfg, prefix, cerr);

    string line;
    while (cin.good()) {
      while (getline(cin, line)) {
        rules.handle(today(), line);
      }
    }

    return EX_OK;
  } catch (missing_rules_order &e) {
    return complain(
      EX_DATAERR
    , format("%1%: no rules.order in '%2%'") % self % inipath
    );
  } catch (exception &e) {
    return complain(
      EX_SOFTWARE
    , format("%1%: internal error: %2%") % self % e.what()
    );
  }
  // let other exceptions cause std::terminate
}
